editor/editor_dock_manager.cpp

```diff
@@ -0,0 +1,712 @@
+						dock_slot[i]->select_next_available();
+						dock_slot[i]->set_block_signals(false);
+					}
+				} else {
+					dock_slot[i]->set_block_signals(true);
+					dock_slot[i]->set_current_tab(first_tab_visible);
+					dock_slot[i]->set_block_signals(false);
+				}
+			} else {
+				dock_slot[i]->hide();
+			}
+		}
+	}
+}
+
+void EditorDockManager::close_all_floating_docks() {
+	for (WindowWrapper *wrapper : floating_docks) {
+		wrapper->set_window_enabled(false);
+	}
+}
+
+void EditorDockManager::add_control_to_dock(DockSlot p_slot, Control *p_control, const String &p_name) {
+	ERR_FAIL_INDEX(p_slot, DOCK_SLOT_MAX);
+	dock_slot[p_slot]->add_child(p_control);
+	if (p_name != "") {
+		dock_slot[p_slot]->set_tab_title(dock_slot[p_slot]->get_tab_idx_from_control(p_control), p_name);
+	}
+}
+
+void EditorDockManager::remove_control_from_dock(Control *p_control) {
+	// If the dock is floating, close it first.
+	for (WindowWrapper *wrapper : floating_docks) {
+		if (p_control == wrapper->get_wrapped_control()) {
+			wrapper->set_window_enabled(false);
+			break;
+		}
+	}
+
+	Control *dock = nullptr;
+	for (int i = 0; i < DOCK_SLOT_MAX; i++) {
+		if (p_control->get_parent() == dock_slot[i]) {
+			dock = dock_slot[i];
+			break;
+		}
+	}
+
+	ERR_FAIL_NULL_MSG(dock, "Control was not in dock.");
+
+	dock->remove_child(p_control);
+	update_dock_slots_visibility();
+}
+
+void EditorDockManager::set_docks_visible(bool p_show) {
+	docks_visible = p_show;
+	update_dock_slots_visibility(true);
+}
+
+bool EditorDockManager::get_docks_visible() const {
+	return docks_visible;
+}
+
+void EditorDockManager::add_vsplit(DockSplitContainer *p_split) {
+	vsplits.push_back(p_split);
+	p_split->connect("dragged", callable_mp(this, &EditorDockManager::_dock_split_dragged));
+}
+
+void EditorDockManager::add_hsplit(DockSplitContainer *p_split) {
+	hsplits.push_back(p_split);
+	p_split->connect("dragged", callable_mp(this, &EditorDockManager::_dock_split_dragged));
+}
+
+void EditorDockManager::register_dock_slot(DockSlot p_dock_slot, TabContainer *p_tab_container) {
+	ERR_FAIL_NULL(p_tab_container);
+	ERR_FAIL_INDEX(p_dock_slot, DOCK_SLOT_MAX);
+
+	dock_slot[p_dock_slot] = p_tab_container;
+
+	p_tab_container->set_custom_minimum_size(Size2(170, 0) * EDSCALE);
+	p_tab_container->set_v_size_flags(Control::SIZE_EXPAND_FILL);
+	p_tab_container->set_popup(dock_select_popup);
+	p_tab_container->connect("pre_popup_pressed", callable_mp(this, &EditorDockManager::_dock_pre_popup).bind(p_dock_slot));
+	p_tab_container->set_drag_to_rearrange_enabled(true);
+	p_tab_container->set_tabs_rearrange_group(1);
+	p_tab_container->connect("tab_changed", callable_mp(this, &EditorDockManager::_dock_tab_changed));
+	p_tab_container->set_use_hidden_tabs_for_min_size(true);
+}
+
+void EditorDockManager::_bind_methods() {
+	ADD_SIGNAL(MethodInfo("layout_changed"));
+}
+
+EditorDockManager::EditorDockManager() {
+	singleton = this;
+
+	dock_select_popup = memnew(PopupPanel);
```

Comment:
This popup could probably be extracted into its own class, it contains plenty of code which doesn't need to be shared.

You also won't need to connect to `theme_changed` this way, you will be able to use the notification.

Reply:
That said, this can probably be done in a follow-up.